### PR TITLE
RES-1234: Add composed lr scheduler

### PR DESCRIPTION
**Pull request by lscheinkman**
@subutai See test and docstring for sample use

### CODE REVIEW

File: tests/unit/frameworks/pytorch/lr_scheduler_test.py [hunk 1]
```diff
@@ -53,5 +53,153 @@ def test_chaining_schedulers(self):
...
+            1.8806,
+            1.5878,
+            1.1786,
+            0.734,
+            0.3421,
+            0.0805,
+        ]
+
+        optimizer = torch.optim.SGD([torch.zeros(1)], lr=1.0)
+        batch_size = 10
```
Comment (subutai) [inline tests/unit/frameworks/pytorch/lr_scheduler_test.py]: I'm confused. Why is steps_per_epoch the same as the batch_size? For clarity it might be better to call this variable `num_batches_per_epoch` or something like that.

File: tests/unit/frameworks/pytorch/lr_scheduler_test.py [hunk 2]
```diff
@@ -53,5 +53,153 @@ def test_chaining_schedulers(self):
...
+        scheduler = ComposedLRScheduler(
+            optimizer,
+            steps_per_epoch=batch_size,
+            schedulers={
+                0: dict(
+                    lr_scheduler_class=OneCycleLR,
+                    lr_scheduler_args=dict(
+                        epochs=10, max_lr=2.0, steps_per_epoch=batch_size
```
Comment (subutai) [inline tests/unit/frameworks/pytorch/lr_scheduler_test.py]: It would be better if the number of epochs and the switch point was not the same number as the steps_per_epoch. For example, epochs could be 10 and steps_per_epoch could be 8.

File: tests/unit/frameworks/pytorch/lr_scheduler_test.py [hunk 3]
```diff
@@ -53,5 +53,153 @@ def test_chaining_schedulers(self):
...
+        expected = [dict(weight_decay=1e-04, momentum=0.0)] * 10
+        expected += [dict(weight_decay=1e-05, momentum=0.9)] * 15
+        scheduler = ComposedLRScheduler(
+            optimizer,
+            steps_per_epoch=batch_size,
+            schedulers={
+                0: dict(
+                    lr_scheduler_class=OneCycleLR,
+                    lr_scheduler_args=dict(
+                        epochs=10,
```
Comment (subutai) [inline tests/unit/frameworks/pytorch/lr_scheduler_test.py]: It would be better if the number of epochs and the switch point was not the same number as the steps_per_epoch.

File: nupic/research/frameworks/pytorch/lr_scheduler.py [hunk 4]
```diff
@@ -43,3 +44,133 @@ def __init__(self, optimizer, lr_scale, last_epoch=-1):
...
+                    cycle_momentum=False,
+                ),
+                optimizer_args=dict(
+                    lr=0.1,
+                    weight_decay=0.0001,
+                    momentum=0.0,
+                    nesterov=False,
+                ),
+            ),
+            35: dict(
```
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/lr_scheduler.py]: Just before. Epoch 35 will use this configuration

[the remark this answers is not in the document]
Comment (subutai) [reply nupic/research/frameworks/pytorch/lr_scheduler.py]: Then I think we want the config before to have epochs=34. We want the full OneCycleLR to happen before the next scheduler takes over.

File: nupic/research/frameworks/pytorch/lr_scheduler.py [hunk 5]
```diff
@@ -43,3 +44,133 @@ def __init__(self, optimizer, lr_scale, last_epoch=-1):
...
+        current_milestone = self.milestones[bisect(self.milestones, current_epoch) - 1]
+
+        # Update LR scheduler and optimizer once the milestone changes
+        if current_batch == 0 and self.active_milestone != current_milestone:
+            self.active_milestone = current_milestone
+            self._update_optimizer()
+            self._update_lr_scheduler()
+        elif isinstance(self.lr_scheduler, OneCycleLR):
+            # Step every batch
+            self.lr_scheduler.step()
```
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/lr_scheduler.py]: it is not. There are 3 different conditions:
1. Is this a new milestone? (current_batch == 0 and self.active_milestone != current_milestone)
2. Is this a batch step? (isinstance(self.lr_scheduler, OneCycleLR))
3. Is this an epoch step? (current_batch == 0 and self.lr_scheduler is not None:)
I can refactor these conditions into nested IF's if it clarifies the intention

File: tests/unit/frameworks/pytorch/lr_scheduler_test.py (hunk 1 above)
Comment (lscheinkman) [reply tests/unit/frameworks/pytorch/lr_scheduler_test.py]: In the context of learning rate schedulers this parameter represent how many times the method `step` will be called per epoch which can be `number of batches` in the case of `OneCycleLR` or `1` for other learning rate schedulers that calls `step` only once per epoch.
`OneCycleLR` also calls this parameter `steps_per_epoch` for the same reason.

File: nupic/research/frameworks/pytorch/lr_scheduler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/lr_scheduler.py]: I'm not sure i understand. 
With this configuration, epochs 0-34(inclusive) will use `OneCycleLR` and `StepLR`  will be used from epoch 35 to the end. Thats how the `MultiStepLR` works as well.
Isn't that what your intention?
